### PR TITLE
Closing issue (#31926) about unknown windows crash when too many arguments given in the config file

### LINKED ISSUE #31926: [Tune]  OSError: [WinError 87] The parameter is incorrect when using pytorch dataloader with num_workers>0 and increasing the config file in windows OS
### Crash when using pytorch dataloader with num_workers>0 and increasing the config file in windows OS

Hi,
I encountered a crash that happens to me. I am still learning how to work with Ray, and experimenting with it to see how it works.
An exception is thrown (attached below) when I add "more things" to the config file that Ray Tune uses.
In my script, I have a Dataloader and the error is thrown from there. The Dataloader has num_workers>0.
In the config file that is created, I have 2 "not_used" variables. If I use only "not_used", It works, and if I use "not_used2" also, it crashes immediately when the tune starts, and one of the actors reaches a point where the Dataloader is used, inside the "train_loop" line "for data in trainloader". 

I attached my script below the exception.

```
2023-01-25 18:37:41,147	ERROR trial_runner.py:1088 -- Trial PytorchTrainable_d55b722d: Error processing event.
ray.exceptions.RayTaskError(OSError): ray::_Inner.train() (pid=80440, ip=127.0.0.1, repr=<ray.tune.trainable.util.PytorchTrainable object at 0x000001FD0F4FF100>)
File "python\ray\_raylet.pyx", line 830, in ray._raylet.execute_task
File "python\ray\_raylet.pyx", line 834, in ray._raylet.execute_task
File "python\ray\_raylet.pyx", line 780, in ray._raylet.execute_task.function_executor
File "C:\Users\sahar\AppData\Roaming\Python\Python39\site-packages\ray\_private\function_manager.py", line 674, in actor_method_executor
return method(__ray_actor, *args, **kwargs)
File "C:\Users\sahar\AppData\Roaming\Python\Python39\site-packages\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
return method(self, *_args, **_kwargs)
File "C:\Users\sahar\AppData\Roaming\Python\Python39\site-packages\ray\tune\trainable\trainable.py", line 367, in train
raise skipped from exception_cause(skipped)
File "C:\Users\sahar\AppData\Roaming\Python\Python39\site-packages\ray\tune\trainable\trainable.py", line 364, in train
result = self.step()
File "C:\Users\sahar\AppData\Roaming\Python\Python39\site-packages\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
return method(self, *_args, **_kwargs)
File "C:\Users\sahar\PycharmProjects\kaggle\facial_keypoints_detection\temp\temp_ray_test2.py", line 187, in step
accuracy, val_loss, val_steps = one_epoch(self.criterion, self.device, self.net,
File "C:\Users\sahar\PycharmProjects\kaggle\facial_keypoints_detection\temp\temp_ray_test2.py", line 102, in one_epoch
train_loop(criterion, device, net, optimizer, trainloader)
File "C:\Users\sahar\PycharmProjects\kaggle\facial_keypoints_detection\temp\temp_ray_test2.py", line 25, in train_loop
for data in trainloader:
File "C:\Users\sahar\PycharmProjects\kaggle\facial_keypoints_detection\venv\lib\site-packages\torch\utils\data\dataloader.py", line 430, in __iter__
self._iterator = self._get_iterator()
File "C:\Users\sahar\PycharmProjects\kaggle\facial_keypoints_detection\venv\lib\site-packages\torch\utils\data\dataloader.py", line 381, in _get_iterator
return _MultiProcessingDataLoaderIter(self)
File "C:\Users\sahar\PycharmProjects\kaggle\facial_keypoints_detection\venv\lib\site-packages\torch\utils\data\dataloader.py", line 1034, in __init__
w.start()
File "C:\Program Files\Python39\lib\multiprocessing\process.py", line 121, in start
self._popen = self._Popen(self)
File "C:\Program Files\Python39\lib\multiprocessing\context.py", line 224, in _Popen
return _default_context.get_context().Process._Popen(process_obj)
File "C:\Program Files\Python39\lib\multiprocessing\context.py", line 327, in _Popen
return Popen(process_obj)
File "C:\Program Files\Python39\lib\multiprocessing\popen_spawn_win32.py", line 73, in __init__
hp, ht, pid, tid = _winapi.CreateProcess(
OSError: [WinError 87] The parameter is incorrect
```

### Versions / Dependencies

```
aiohttp==3.8.3
aiohttp-cors==0.7.0
aiorwlock==1.3.0
aiosignal==1.3.1
ansicon==1.89.0
anyio==3.6.2
argcomplete==2.0.0
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
arrow==1.2.3
asttokens==2.2.1
async-timeout==4.0.2        
attrs==22.2.0
backcall==0.2.0
beautifulsoup4==4.11.1      
bleach==5.0.1
blessed==1.19.1
boto==2.49.0
cachetools==5.3.0
certifi==2022.12.7
cffi==1.15.0
charset-normalizer==2.1.1
click==8.1.3
clip==1.0
cloudpickle==2.2.1
colorama==0.4.6
colorful==0.5.5
comm==0.1.2
contourpy==1.0.6
crcmod==1.7
cryptography==36.0.1
cycler==0.11.0
Cython==0.29.24
debugpy==1.6.5
decorator==5.1.1
defusedxml==0.7.1
distlib==0.3.6
dm-tree==0.1.8
entrypoints==0.4
executing==1.2.0
fastapi==0.89.1
fasteners==0.17.3
fastjsonschema==2.16.2
filelock==3.9.0
Flask==2.2.2
fonttools==4.38.0
fqdn==1.5.1
frozenlist==1.3.3
fsspec==2023.1.0
ftfy==6.1.1
future==0.18.3
gcs-oauth2-boto-plugin==3.0
google-api-core==2.11.0
google-apitools==0.5.32
google-auth==2.16.0
google-reauth==0.1.1
googleapis-common-protos==1.58.0
gpustat==1.0.0
grpcio==1.51.1
gsutil==5.6
gym==0.23.1
gym-notices==0.0.8
h11==0.14.0
httplib2==0.20.2
hyperopt==0.2.7
idna==3.4
imageio==2.25.0
importlib-metadata==6.0.0
ipykernel==6.20.1
ipython==8.8.0
ipython-genutils==0.2.0
ipywidgets==8.0.4
isoduration==20.11.0
itsdangerous==2.1.2
jedi==0.18.2
Jinja2==3.1.2
jinxed==1.2.0
joblib==1.2.0
jsonpointer==2.3
jsonschema==4.17.3
jupyter==1.0.0
jupyter-console==6.4.4
jupyter-events==0.6.3
jupyter_client==7.4.9
jupyter_core==5.1.3
jupyter_server==2.1.0
jupyter_server_terminals==0.4.4
jupyterlab-pygments==0.2.2
jupyterlab-widgets==3.0.5
keras==2.7.0
kiwisolver==1.4.4
lz4==4.3.2
markdown-it-py==2.1.0
MarkupSafe==2.1.1
matplotlib==3.6.3
matplotlib-inline==0.1.6
mdurl==0.1.2
mistune==2.0.4
monotonic==1.6
msgpack==1.0.4
multidict==6.0.4
nbclassic==0.4.8
nbclient==0.7.2
nbconvert==7.2.7
nbformat==5.7.3
nest-asyncio==1.5.6
networkx==3.0
notebook==6.5.2
notebook_shim==0.2.2
numpy==1.24.1
nvidia-ml-py==11.495.46
oauth2client==4.1.3
opencensus==0.11.1
opencensus-context==0.1.3
opencv-python==4.7.0.68
packaging==23.0
pandas==1.5.2
pandocfilters==1.5.0
parso==0.8.3
pickleshare==0.7.5
Pillow==9.4.0
platformdirs==2.6.2
prometheus-client==0.13.1
prompt-toolkit==3.0.36
protobuf==3.20.1
psutil==5.9.4
pure-eval==0.2.2
py-spy==0.3.14
py4j==0.10.9.7
pyarrow==6.0.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.21
pydantic==1.10.4
Pygments==2.14.0
pyOpenSSL==22.0.0
pyparsing==3.0.7
pyrsistent==0.19.3
python-dateutil==2.8.2
python-json-logger==2.0.4
pytz==2022.7
pyu2f==0.1.5
PyWavelets==1.4.1
pywin32==305
pywinpty==2.0.10
PyYAML==6.0
pyzmq==25.0.0
qtconsole==5.4.0
QtPy==2.3.0
ray==2.2.0
regex==2022.10.31
requests==2.28.2
retry-decorator==1.1.1
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rich==13.2.0
rsa==4.9
scikit-image==0.19.3
scikit-learn==1.2.0
scipy==1.10.0
Send2Trash==1.8.0
six==1.16.0
smart-open==6.3.0
sniffio==1.3.0
soupsieve==2.3.2.post1
stack-data==0.6.2
starlette==0.22.0
tabulate==0.9.0
tensorboardX==2.5.1
terminado==0.17.1
threadpoolctl==3.1.0
tifffile==2023.1.23.1
tinycss2==1.2.1
torch==1.13.1+cu117
torchaudio==0.13.1+cu117
torchsummary==1.5.1
torchvision==0.14.1+cu117
tornado==6.2
tqdm==4.64.1
traitlets==5.8.1
typer==0.7.0
typing_extensions==4.4.0
uri-template==1.2.0
urllib3==1.26.14
uvicorn==0.20.0
vecsim==0.0.51
virtualenv==20.17.1
wcwidth==0.2.6
webcolors==1.12
webencodings==0.5.1
websocket-client==1.4.2
Werkzeug==2.2.2
widgetsnbextension==4.0.5
yarl==1.8.2
zipp==3.11.0
```

### Reproduction script

```
import os
from typing import Union, Dict, Optional

import ray
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from filelock import FileLock
from hyperopt import hp
from ray import tune
from ray.tune import CLIReporter
from ray.tune.experiment.trial import ExportFormat
from ray.tune.schedulers import ASHAScheduler
from ray.tune.search.hyperopt import HyperOptSearch
from torch.utils.data import random_split

def train_loop(criterion, device, net, optimizer, trainloader):
    running_loss = 0.0
    epoch_steps = 0
    net.train()
    for data in trainloader:
        # get the inputs; data is a list of [inputs, labels]
        inputs, labels = data
        inputs, labels = inputs.to(device), labels.to(device)

        # zero the parameter gradients
        optimizer.zero_grad()

        # forward + backward + optimize
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        # print statistics
        running_loss += loss.item()
        epoch_steps += 1


def val_loop(criterion, device, net, valloader):
    # Validation loss
    val_loss = 0.0
    val_steps = 0
    total = 0
    correct = 0
    net.eval()
    for data in valloader:
        with torch.no_grad():
            inputs, labels = data
            inputs, labels = inputs.to(device), labels.to(device)

            outputs = net(inputs)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

            loss = criterion(outputs, labels)
            val_loss += loss.cpu().numpy()
            val_steps += 1
    accuracy = correct / total
    return accuracy, val_loss, val_steps


class Net(nn.Module):
    def __init__(self, l1=120, l2=84, c1=6, c2=16, k1=5,
                 batch_norms=False,
                 dropout_rate=0.5
                 ):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, c1, k1)
        self.conv2 = nn.Conv2d(c1, c2, 3)
        self.conv3 = nn.Conv2d(c2, 32, 3)
        self.fc1 = nn.LazyLinear(l1)
        self.fc2 = nn.Linear(l1, l2)

        self.fc3 = nn.Linear(l2, 10)

        self.pool = nn.MaxPool2d(2, 2)
        self.flatten = nn.Flatten()
        self.drop_out = nn.Dropout(dropout_rate)
        self.batch_norm1 = nn.BatchNorm2d(c1)
        # self.batch_norm1 = nn.BatchNorm2d(c1) if batch_norm else nn.Identity()
        self.batch_norm2 = nn.BatchNorm2d(c2)
        # self.batch_norm2 = nn.BatchNorm2d(c2) if batch_norm else nn.Identity()

    def forward(self, x):
        x = self.drop_out(self.pool(F.relu(self.batch_norm1(self.conv1(x)))))
        x = self.drop_out(self.pool(F.relu(self.batch_norm2(self.conv2(x)))))
        x = self.drop_out(self.pool(F.relu(self.conv3(x))))
        x = self.flatten(x)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x


def one_epoch(criterion, device, net, optimizer, trainloader, valloader):
    train_loop(criterion, device, net, optimizer, trainloader)
    accuracy, val_loss, val_steps = val_loop(criterion, device, net, valloader)
    return accuracy, val_loss, val_steps


def report_epoch(accuracy, epoch, net, optimizer, time_passed_for_epoch_mean, val_loss, val_steps):
    if ray.tune.is_session_enabled():
        with tune.checkpoint_dir(epoch) as checkpoint_dir:
            path = os.path.join(checkpoint_dir, "checkpoint")
            torch.save((net.state_dict(), optimizer.state_dict()), path)
        tune.report(loss=(val_loss / val_steps), accuracy=accuracy,
                    time_passed_for_epoch_mean=time_passed_for_epoch_mean)
    else:
        print("Validation loss: %.3e, accuracy: %.3f time_passed_for_epoch: %.3f" % (
            val_loss / val_steps, accuracy, time_passed_for_epoch_mean))


def setup_for_train(config):
    net = Net(config["l1"],
              config["l2"],
              config["c1"],
              config["c2"],
              config["k1"],
              # config["batch_norms"],
              config["dropout_rate"]
              )
    device = "cpu"
    if torch.cuda.is_available():
        device = "cuda:0"
        if torch.cuda.device_count() > 1:
            net = nn.DataParallel(net)
    net.to(device)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(net.parameters(), lr=config["lr"])
    return criterion, device, net, optimizer


def load_data(data_dir="./data"):
    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
    ])

    with FileLock(os.path.expanduser(data_dir + ".lock")):
        trainset = torchvision.datasets.CIFAR10(
            root=data_dir, train=True, download=True, transform=transform)

        testset = torchvision.datasets.CIFAR10(
            root=data_dir, train=False, download=True, transform=transform)

    return trainset, testset


def get_dataloaders(config, data_dir):
    trainset, testset = load_data(data_dir)
    test_abs = int(len(trainset) * 0.8)
    train_subset, val_subset = random_split(
        trainset, [test_abs, len(trainset) - test_abs])
    trainloader = torch.utils.data.DataLoader(
        train_subset,
        batch_size=int(config["batch_size"]),
        shuffle=True,
        prefetch_factor=2, num_workers=3,
        persistent_workers=True,
        # pin_memory=True
    )
    valloader = torch.utils.data.DataLoader(
        val_subset,
        batch_size=int(config["batch_size"]),
        shuffle=True,
        prefetch_factor=2, num_workers=3,
        persistent_workers=True,
        # pin_memory=True
    )
    return trainloader, valloader


class PytorchTrainable(tune.Trainable):
    def setup(self, config, checkpoint_dir=None, data_dir=None):
        self.not_used = config["not_used"]
        self.not_used2 = config["not_used2"]
        self.trainloader, self.valloader = get_dataloaders(config, data_dir)
        self.criterion, self.device, self.net, self.optimizer = setup_for_train(self.config)

    def step(self):
        accuracy, val_loss, val_steps = one_epoch(self.criterion, self.device, self.net,
                                                  self.optimizer, self.trainloader, self.valloader)
        return {
            "loss": (val_loss / val_steps),
            "accuracy": accuracy,
        }

    def reset_config(self, new_config: Dict) -> bool:
        self.config = new_config
        self.criterion, self.device, self.net, self.optimizer = setup_for_train(self.config)
        return True

    def _export_model(self, export_formats, export_dir):
        ExportFormat.validate(export_formats)
        exported = {}
        if ExportFormat.MODEL in export_formats:
            path = os.path.join(export_dir, ExportFormat.MODEL)
            torch.save(self.net.state_dict(), path)
            exported[ExportFormat.MODEL] = path
        return exported

    def cleanup(self):
        del self.net
        torch.cuda.empty_cache()

    def save_checkpoint(self, checkpoint_dir: str) -> Optional[Union[str, Dict]]:
        path = os.path.join(checkpoint_dir, "checkpoint")
        torch.save((self.net.state_dict(), self.optimizer.state_dict()), path)
        return path

    def load_checkpoint(self, checkpoint: Union[Dict, str]):
        if isinstance(checkpoint, str):
            checkpoint = torch.load(checkpoint)
        self.net.load_state_dict(checkpoint[0])
        self.optimizer.load_state_dict(checkpoint[1])


def main(num_samples=10, grace_period=1, reduction_factor=4, cpu_per_trial=2, max_concurrent_trials=5,
         gpus_per_trial=2.0, max_num_epochs=10):
    data_dir = os.path.abspath("./data")
    load_data(data_dir)
    config = get_config()

    scheduler = get_schedualer(grace_period, max_num_epochs, reduction_factor)
    reporter = get_reporter(config)
    # ray.init(local_mode=True)
    hyperopt_search = get_hyper_opt_search(config)
    tune_run_result = tune.run(
        tune.with_parameters(PytorchTrainable, checkpoint_dir=None, data_dir=data_dir),
        resources_per_trial={"cpu": cpu_per_trial, "gpu": gpus_per_trial},
        config=config,
        num_samples=num_samples,
        scheduler=scheduler,
        progress_reporter=reporter,
        local_dir="./logs",
        fail_fast=True,
        max_concurrent_trials=max_concurrent_trials,
        search_alg=hyperopt_search,
        verbose=1,
        reuse_actors=True,
    )


def get_config():
    return {
        "epochs": 10,
        "not_used": 5,
        "not_used2": 5,
        "lr": hp.uniform("lr", 1e-4, 1e-1),
        "l1": hp.choice("l1", [2 ** n for n in range(4, 11)]),
        "l2": hp.choice("l2", [2 ** n for n in range(4, 11)]),
        "c1": hp.choice("c1", [2 ** n for n in range(3, 9)]),
        "c2": hp.choice("c2", [2 ** n for n in range(3, 9)]),
        "k1": hp.choice("k1", [3, 5, 7]),
        "dropout_rate": hp.choice("dropout_rate", [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7]),
        "batch_size": hp.choice("batch_size", [64, 128, 256])
    }


def get_hyper_opt_search(config):
    return HyperOptSearch(config, metric="loss", mode="min")


def get_reporter(config):
    reporter = CLIReporter(
        parameter_columns=sorted(list(config.keys())),
        metric_columns=["loss", "accuracy", "training_iteration", "time_passed_for_epoch_mean"],
        sort_by_metric=True,
        metric="loss",
        mode="min"
    )
    return reporter


def get_schedualer(grace_period, max_num_epochs, reduction_factor):
    scheduler = ASHAScheduler(
        metric="loss",
        mode="min",
        max_t=max_num_epochs,
        grace_period=grace_period,
        reduction_factor=reduction_factor)
    return scheduler


if __name__ == "__main__":
    max_concurrent_trials = 5
    main(num_samples=10, grace_period=4, reduction_factor=2, cpu_per_trial=4,
         max_concurrent_trials=max_concurrent_trials,
         gpus_per_trial=1 / max_concurrent_trials, max_num_epochs=25)

```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by Sahar-E**
## Why are these changes needed?

There is a crash that I encountered in Windows. It related to the fact that the path was too long for windows. 
So to allow the user to be aware of this issue, I added a check in the code that checks if the path is 
too long and warn it with a logger warning message.

## Related issue number

Closes: #31926.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested

### CODE REVIEW

File: python/ray/constants.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# OS names:
+WINDOWS_OS = "Windows"
```
Comment (jjyao) [inline python/ray/constants.py:2]: we already have `python/ray/_private/ray_constants.py`.

For this case, I actually think it's fine to use the string literals directly. These string literals will not change.

Comment (Sahar-E) [reply python/ray/constants.py:2]: I agree, thanks.

File: python/ray/tune/experiment/trial.py [hunk 2]
```diff
@@ -641,10 +641,36 @@ def init_logdir(self):
...
+        """
+        safety_buffer = 15  # arbitrary size to account for other path components.
+        if platform.system() == "Windows":
+            # Windows has a 260-character limit:
+            return len(str(logdir_path)) + safety_buffer > 260
+        elif platform.system() == "Linux":
+            # Linux has a 4096-character limit:
+            return len(str(logdir_path)) + safety_buffer > 4096
+        return False
+
```
Comment (krfricke) [inline python/ray/tune/experiment/trial.py:673]: Since this is a static method, can we just move it outside the class?

**Review by krfricke — CHANGES_REQUESTED** [review]
Thanks! Minor nit, let's move this outside of the `Trial` class (it can still remain in the same file)

**Comment by Sahar-E** [comment]
Hi @krfricke @jjyao,
Do you know why "buildkite/oss-ci-build-pr/book-linkcheck — Soft failed" fails and how can I fix it?
Thanks,
Sahar

**Review by Sahar-E — COMMENTED** [review]
.

**Review by Sahar-E — COMMENTED** [review]
.

**Review by krfricke — APPROVED** [review]
This looks good to me! Thanks for the contribution and sorry for the delay in processing

**Comment by krfricke** [comment]
I've changed the logic a bit to be more OS-agnostic. Let me know if you think I'm missing something there!

File: python/ray/tune/experiment/trial.py [hunk 3]
```diff
@@ -799,13 +799,15 @@ def init_local_path(self):
             self.relative_logdir = _create_unique_logdir_name(
                 str(self.local_experiment_path), self._generate_dirname()
             )
-        assert self.logdir
-        logdir_path = Path(self.logdir)
-        if _is_path_too_long(logdir_path):
+        assert self.local_path
+        logdir_path = Path(self.local_path)
+        max_path_length = os.pathconf("/", "PC_PATH_MAX")
```
Comment (Sahar-E) [inline python/ray/tune/experiment/trial.py:804]: I think this function is not available on Windows.

**Review by Sahar-E — COMMENTED** [review]
Thanks @krfricke,
I think pathconf is not available in Windows.

Sahar

File: python/ray/tune/experiment/trial.py (hunk 3 above)
Comment (krfricke) [reply python/ray/tune/experiment/trial.py]: I see - re-introduced a function for this

**Pull request merged**